### PR TITLE
TEP-0061 Allow custom task to be embedded in pipeline, proposal.

### PR DESCRIPTION
**Pull request by ScrapCodes**
Added details for proposal and an example.

One of the consideration of this proposal is: https://github.com/tektoncd/experimental/pull/730#discussion_r615896845

/cc @Tomcli @ImJasonH @vdemeester @sbwsg @pritidesai @afrittoli

### CODE REVIEW

**Comment by ScrapCodes** [comment]
Hello @pritidesai and @sbwsg , Thank you for assigning yourself, if this TEP proposal looks good, then I can work on the design and implementation. Please let me know, if this makes sense and if there are any processes need to be followed for the approval of the proposed API.

Thank you again !

File: teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md [hunk 1]
```diff
@@ -112,9 +112,177 @@ be handled, or user scenarios that will be affected and must be accomodated.
...
+               script: |
+                 #!/usr/bin/env bash
+                 echo "I am the last task after the loop task"
+```
+
+`Tektoncd/pipeline` can only validate the structure and fields it knows about, validation of the
+custom task spec field(s) is delegated to the custom task controller.
+
+A custom controller may still choose to not support one of `Spec` based `Run` specification or
+`Ref` based ones. This can be done by implementing suitable webhooks at the custom controller end.
```
Comment (sbwsg) [inline teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md:280]: Just to clarify this a bit more: what kind of effect does this have on the PipelineRun?  Would it have to wait for its timeout deadline or would it be able to immediately detect the error and be placed into a Succeeded/False condition?

Comment (Tomcli) [reply teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md:280]: Currently with taskRef, if the custom task controller did not reconcile the `Run` CRD, then PipelineRun just keep waiting  until its timeout deadline. We will follow the same behavior for taskSpec (Run Spec) as well.

@ScrapCodes can you update the doc to clarify this behavior? thanks

Comment (ScrapCodes) [reply teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md]: I just updated the doc.

**Review by sbwsg — APPROVED** [review]

**Comment by ScrapCodes** [comment]
@sbwsg Thank you very much Scott, for approving the PR.

**Comment by ScrapCodes** [comment]
Hi @pritidesai , we are just one `approve` away from getting this going forward. Can you please take a look?

File: teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md [hunk 2]
```diff
@@ -112,9 +112,186 @@ be handled, or user scenarios that will be affected and must be accomodated.
...
+
+	// +optional
+	Spec runtime.RawExtension `json:"spec,omitempty"`
+
+	// +optional
+	Metadata PipelineTaskMetadata `json:"metadata,omitempty"`
+
+	// TaskSpec is a specification of a task
+	TaskSpec `json:",inline,omitempty"`
+}
```
Comment (pritidesai) [inline teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md:174]: Thank you for adding these details. This is how the new `struct` would look like. 

This is what we have today:

https://github.com/tektoncd/pipeline/blob/3dc52df0d5571b7839929d49ae23e5bb6a11f101/pkg/apis/pipeline/v1beta1/pipeline_types.go#L111-L117

File: teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md [hunk 3]
```diff
@@ -112,9 +112,186 @@ be handled, or user scenarios that will be affected and must be accomodated.
...
+
+```go
+// RunSpec defines the desired state of Run
+type RunSpec struct {
+	// +optional
+	Ref *TaskRef `json:"ref,omitempty"`
+
+	// TaskSpec is a specification of a custom task
+	// +optional
+	Spec *v1beta1.EmbeddedTask `json:"spec,omitempty"`
```
Comment (pritidesai) [inline teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md:134]: 👍  new field being added

File: teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md [hunk 4]
```diff
@@ -112,9 +112,186 @@ be handled, or user scenarios that will be affected and must be accomodated.
...
+
+A custom controller may still choose to not support one of `Spec` or `Ref` based 
+`Run` or `PipelineRun` specification. This can be done by implementing validations at the custom controller end.
+If the custom controller did not respond in any of the ways i.e. either validation errors or reconcile CRD,
+then, a `PipelineRun` or a `Run` will wait until the timeout and mark the status as `Failed`.
+
+What is the fate of an existing custom controller developed prior to the implementation of this TEP. If the 
+custom controller implemented a validation for missing a `Ref`, then the `PipelineRun` or `Run` 
+missing a `Ref` will fail immediately with configured error and if however, no validation was
+implemented for missing a `Ref`, then it can even lead to `nil dereference errors` or have the same fate
```
Comment (pritidesai) [inline teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md:288]: thank you for documenting this here. We highly encourage the custom controller to have some kind of validation on either one of the fields being initialized `ref` or `spec`.

**Comment by pritidesai** [comment]
thank you @ScrapCodes and @Tomcli for the great work 👍 

/lgtm

Please add some more details on the tekton webhook validation in next PR when you move this TEP to `implementable`. 

While validating `pipelineTask`, `validateCustomTask` will be updated to include either `taskRef` or `taskSpec` is specified. And if `taskSpec` is specified, `APIVersion` and `Kind` both exist.

https://github.com/tektoncd/pipeline/blob/3dc52df0d5571b7839929d49ae23e5bb6a11f101/pkg/apis/pipeline/v1beta1/pipeline_types.go#L187-L190

Also, for a non custom task, these new fields are not allowed.

**Comment by ScrapCodes** [comment]
Thank you @Tomcli and @pritidesai for the impetus for this TEP. Especially for the clear guidance above and all along in the development of this TEP.  😄

**Pull request merged**